### ISSUE TITLE
How to use "vl_nnsoftmax" function in network training?

### ISSUE THREAD

**Issue by luzhi** [vlfeat/matconvnet#166]
I found the last layer of all the examples in Matconvnet is "**vl_nnsoftmaxloss**". But I need to evaluate the network by the "probability" values of each sample that belongs to a particular class. So, I tried to plug the "**vl_nnsoftmax**" function in the last layer. But the system always shows an error message liek this:

> training: epoch 01: batch   1/ 12: Error using  .* 
> Matrix dimensions must agree.

> Error in vl_nnsoftmax (line 30)
> Y = Y .* bsxfun(@minus, dzdY, sum(dzdY .* Y, 3)) ;

> Error in vl_simplenn (line 283)
>         res(i).dzdx = vl_nnsoftmax(res(i).x, res(i+1).dzdx) ;

> Error in zl_cnn_train>process_epoch (line 311)
>    res = vl_simplenn(net, im, dzdy, res, ...

> Error in zl_cnn_train (line 168)
>    [net,stats.train] = process_epoch(opts, getBatch, epoch, train, learningRate, imdb, net) ;

> Error in zlscript_ROI_simpleCNN (line 61)
> [net, info] = cnn_train(net, imdb, @getBatch, ...

How can I use the "**vl_nnsoftmax**" function? Thanks!

**Comment by mingtop** [comment]
1 * 1 * classsed * numbers 

Do yours labels assemble in this way ? 

Check the example of mnist


imdb.images.data = data ;
imdb.images.data_mean = dataMean;
imdb.images.labels = cat(2, y1, y2) ;
imdb.images.set = set ;
imdb.meta.sets = {'train', 'val', 'test'} ;
imdb.meta.classes = arrayfun(@(x)sprintf('%d',x),0:9,'uniformoutput',false) ;


-----------------------

**Comment by vedaldi** [comment]
Hi, it should be 1 x 1 x 1 x N unless you want to specify a different label per location. See also here: http://www.vlfeat.org/matconvnet/mfiles/vl_nnsoftmaxloss/ <http://www.vlfeat.org/matconvnet/mfiles/vl_nnsoftmaxloss/>

**Comment by jshtok** [comment]
Hi Andrea,

First of all, thank you for making the effort to build this Deep LEarning package in Matlab. I haven't mastered it yet, but after struggling with Caffe, its like someone finally switched the lights on :)

To the point, I am trying to fine-tune the classifier imagenet-vgg-f on my data. The last layer is 'softmax'. As I start training, the initial value for dzdy gradient is set in code (cnn_train.m, line 286) to be one;  however, when the vl_nnsoftmax is executed (vl_simplenn.m, line 359) the data input has the dimension [1x1x1000x68] (1k classes, 68 examples), and this doesn't match the dzdY dimension. 
Can you please help me resolve this? 

Thank you,
Joseph

**Comment by omegafil** [comment]
Hello,

actually the function that he's trying to use is [softmax] (http://www.vlfeat.org/matconvnet/mfiles/vl_nnsoftmax/) , isnt'it?

**Comment by angy50** [comment]
hi
I am facing the same problem. It is due to mismatch of dimensions of input "desired class C" of softmaxloss layer during training and testing. During training class of softmaxloss layer(SMLL) takes the size of number of images in last trained batch(for e.g. if in last batch of training, there are 100 images, size will be 1 by 100) indicating one label per image but while testing trained net on a single image, SMLL takes 1 by 100 as desired class dimension ehereas expected class dimension is only 1 which creates a mismatch. 
please suggest any solution for it.
thank you

**Comment by micklexqg** [comment]
@luzhi , have you find the method to to evaluate the network by the "probability" values of each sample that belongs to a particular class? I also need the  "probability" .
